File: beginner/091/D.py
from bisect import bisect_left, bisect_right
N = int( input())
A = list( map( int, input().split()))
B = list( map( int, input().split()))
ans = 0
for k in range(29):
    T = 2**k
    S = 2**(k+1)
    kB = list( map( lambda x:x%S, B))
    kB.sort()
    check = 0
    for i in range(N):
        a = A[i]%S
        check += bisect_left(kB, (2*T-a))- bisect_left(kB, (T-a))
        check += bisect_left(kB, (4*T-a))- bisect_left(kB, (3*T-a))
        check %= 2
    if check == 1:
        ans += 2**k
print(ans)

File: beginner/091/D.cpp
#include <iostream>
#include <vector>
#include <cstdio>
#include <string>
#include <algorithm>
#include <climits>
#include <math.h>
using namespace std;
typedef long long int ll;
#define all(x) x.begin(),x.end()
int main()
{
  int N;
  ll a, t, now;
  int p;
  cin >> N;
  vector<int> vecA;
  vector<int> vecB;
  for (int i = 0; i < N; ++i) {
    cin >> a;
    vecA.push_back(a);
  }
  for (int i = 0; i < N; ++i) {
    cin >> a;
    vecB.push_back(a);
  }
  vector<int> check(N);
  int ans[30];
  for (int i = 0; i < 30; ++i) {
    now = 0;
    t = pow(2,i);
    for (int j = 0; j < N; ++j) {
      check[j] = vecA[j]%(2*t);
    }
    sort(all(check));
    for (int j = 0; j < N; ++j) {
      a = vecB[j]%(t*2);
      now += (lower_bound(all(check),t*2-a) - lower_bound(all(check), t-a)) + (lower_bound(all(check), 4*t-a) - lower_bound(all(check), t*3-a));
    }
    ans[i] = now%2;
  }
  ll ret = 0;
  for (int i = 0; i < 30; ++i) {
    ret += ans[i]*pow(2,i);
  }
  cout << ret << "\n";
  return 0;
}
